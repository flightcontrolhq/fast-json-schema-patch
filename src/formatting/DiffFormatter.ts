import { resolvePatchPath } from "../utils/pathUtils";
import { cachedJsonStringify, cachedBuildPathMap } from "../performance/cache";
import { fastHash } from "../performance/fashHash";
import type {
  DiffLine,
  JsonValue,
  JsonObject,
  Operation,
  PathMap,
  SideBySideDiff,
  UnifiedDiffLine,
} from "../types";
import type { ArrayPlan } from "../core/buildPlan";

// Enhanced caching for diff formatters with content-based keys
const diffFormatterCache = new Map<string, SideBySideDiff>();

function getPathLineRange(
  pathMap: PathMap,
  path: string,
  jsonObj: JsonValue,
  isForNewVersion = false
): { start: number; end: number } | null {
  const resolvedPath = resolvePatchPath(path, jsonObj, isForNewVersion);
  if (!resolvedPath) return null;

  const info = pathMap[resolvedPath];
  if (info?.value && info.valueEnd) {
    return {
      start: info.value.line + 1,
      end: info.valueEnd.line + 1,
    };
  }

  const pathParts = resolvedPath.split("/").filter((p) => p !== "");
  for (let i = pathParts.length; i > 0; i--) {
    const parentPath = `/${pathParts.slice(0, i).join("/")}`;
    const parentInfo = pathMap[parentPath];
    if (parentInfo?.value && parentInfo.valueEnd) {
      return {
        start: parentInfo.value.line + 1,
        end: parentInfo.valueEnd.line + 1,
      };
    }
  }
  return null;
}

export class DiffFormatter {
  private originalJson: JsonValue;
  private newJson: JsonValue;
  private originalPathMap: PathMap;
  private newPathMap: PathMap;
  private plan?: ArrayPlan;

  constructor(originalJson: JsonValue, newJson: JsonValue, plan?: ArrayPlan) {
    this.originalJson = originalJson;
    this.newJson = newJson;
    this.originalPathMap = cachedBuildPathMap(originalJson);
    this.newPathMap = cachedBuildPathMap(newJson);
    this.plan = plan;
  }

  format(patches: Operation[]): SideBySideDiff {
    // Enhanced caching: create a cache key based on patches and plan
    const patchesKey = this.createPatchesKey(patches);
    const planKey = this.plan
      ? `${this.plan.primaryKey || ""}-${
          this.plan.hashFields?.join(",") || ""
        }-${this.plan.strategy || ""}`
      : "default";
    const contentHash = fastHash(
      {
        original: JSON.stringify(this.originalJson).substring(0, 100),
        new: JSON.stringify(this.newJson).substring(0, 100),
      },
      ["original", "new"]
    );
    const cacheKey = `${contentHash}-${patchesKey}-${planKey}`;

    // Check cache first
    const cached = diffFormatterCache.get(cacheKey);
    if (cached) {
      return cached;
    }

    // Generate the diff
    const result = this.generateDiff(patches);

    // Cache the result (keep cache size reasonable)
    if (diffFormatterCache.size > 1000) {
      // Simple cache eviction - clear half when full
      const keys = Array.from(diffFormatterCache.keys());
      for (let i = 0; i < keys.length / 2; i++) {
        diffFormatterCache.delete(keys[i] as string);
      }
    }
    diffFormatterCache.set(cacheKey, result);

    return result;
  }

  private createPatchesKey(patches: Operation[]): string {
    // Create a lightweight hash of the patches for caching
    if (patches.length === 0) return "empty";

    // Use enhanced hashing for consistent cache keys
    const patchData = {
      count: patches.length,
      operations: patches.map((p) => `${p.op}:${p.path}`).join(","),
      // Include a sample of patch content for uniqueness
      sample: patches
        .slice(0, 3)
        .map((p) => p.op)
        .join(""),
    };

    return fastHash(patchData as JsonObject, ["count", "operations", "sample"]);
  }

  private generateDiff(patches: Operation[]): SideBySideDiff {
    const originalAffectedLines = new Set<number>();
    const newAffectedLines = new Set<number>();

    // Enhanced patch processing with schema awareness
    for (const op of patches) {
      if (op.op === "remove" || op.op === "replace") {
        const range = getPathLineRange(
          this.originalPathMap,
          op.path,
          this.originalJson,
          false
        );
        if (range) {
          for (let i = range.start; i <= range.end; i++) {
            originalAffectedLines.add(i);
          }
        }
      }

      if (op.op === "add" || op.op === "replace") {
        const range = getPathLineRange(
          this.newPathMap,
          op.path,
          this.newJson,
          true
        );
        if (range) {
          for (let i = range.start; i <= range.end; i++) {
            newAffectedLines.add(i);
          }
        }
      }
    }

    // Use cached JSON stringification
    const originalFormatted = cachedJsonStringify(this.originalJson);
    const newFormatted = cachedJsonStringify(this.newJson);

    const originalLines = originalFormatted.split("\n");
    const newLines = newFormatted.split("\n");

    const originalDiffLines: DiffLine[] = originalLines.map((line, index) => ({
      lineNumber: index + 1,
      content: line,
      type: originalAffectedLines.has(index + 1) ? "removed" : "unchanged",
    }));

    const newDiffLines: DiffLine[] = newLines.map((line, index) => ({
      lineNumber: index + 1,
      content: line,
      type: newAffectedLines.has(index + 1) ? "added" : "unchanged",
    }));

    // Enhanced unified diff generation
    const unified = this.generateUnifiedDiff(originalDiffLines, newDiffLines);

    return {
      originalLines: originalDiffLines,
      newLines: newDiffLines,
      unifiedDiffLines: unified,
    };
  }

  private generateUnifiedDiff(
    originalDiffLines: DiffLine[],
    newDiffLines: DiffLine[]
  ): UnifiedDiffLine[] {
    const unified: UnifiedDiffLine[] = [];
    let i = 0;
    let j = 0;

    while (i < originalDiffLines.length || j < newDiffLines.length) {
      const iLine = originalDiffLines[i];
      const jLine = newDiffLines[j];

      if (iLine?.type === "unchanged" && jLine?.type === "unchanged") {
        if (iLine && jLine) {
          unified.push({
            type: "unchanged",
            content: iLine.content,
            oldLineNumber: iLine.lineNumber,
            newLineNumber: jLine.lineNumber,
            key: `unchanged-${iLine.lineNumber}-${jLine.lineNumber}`,
          });
        }
        i++;
        j++;
        continue;
      }

      const iBefore = i;
      const jBefore = j;

      while (
        i < originalDiffLines.length &&
        originalDiffLines[i]?.type === "removed"
      ) {
        const line = originalDiffLines[i];
        if (line) {
          unified.push({
            type: "removed",
            content: line.content,
            oldLineNumber: line.lineNumber,
            key: `removed-${line.lineNumber}`,
          });
        }
        i++;
      }

      while (j < newDiffLines.length && newDiffLines[j]?.type === "added") {
        const line = newDiffLines[j];
        if (line) {
          unified.push({
            type: "added",
            content: line.content,
            newLineNumber: line.lineNumber,
            key: `added-${line.lineNumber}`,
          });
        }
        j++;
      }

      // If we're stuck, advance pointers to avoid infinite loop
      if (i === iBefore && j === jBefore) {
        if (i < originalDiffLines.length) {
          const line = originalDiffLines[i];
          if (line) {
            unified.push({
              type: line.type,
              content: line.content,
              oldLineNumber: line.lineNumber,
              key: `stuck-i-${line.lineNumber}`,
            });
          }
          i++;
        }
        if (j < newDiffLines.length) {
          const line = newDiffLines[j];
          if (line) {
            unified.push({
              type: line.type,
              content: line.content,
              newLineNumber: line.lineNumber,
              key: `stuck-j-${line.lineNumber}`,
            });
          }
          j++;
        }
      }
    }

    return unified;
  }
}
