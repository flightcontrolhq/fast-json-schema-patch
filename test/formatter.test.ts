import { describe, test, expect } from "bun:test";
import { buildPlan, JsonSchemaPatcher } from "../src/index";
import { DiffFormatter } from "../src/formatting/DiffFormatter";
import { faker } from "@faker-js/faker";

const userSchema = {
  type: "object",
  properties: {
    userId: { type: "string" },
    username: { type: "string" },
    email: { type: "string" },
    posts: {
      type: "array",
      items: {
        type: "object",
        properties: {
          postId: { type: "string" },
          title: { type: "string" },
          content: { type: "string" },
          tags: { type: "array", items: { type: "string" } },
        },
        required: ["postId", "title", "content"],
      },
    },
    metadata: {
      type: "object",
      additionalProperties: true,
    },
  },
  required: ["userId", "username", "email"],
};

function createRandomUser() {
  return {
    userId: faker.string.uuid(),
    username: faker.internet.username(),
    email: faker.internet.email(),
    posts: Array.from(
      { length: faker.number.int({ min: 2, max: 3 }) },
      () => ({
        postId: faker.string.uuid(),
        title: faker.lorem.sentence(),
        content: faker.lorem.paragraphs(),
        tags: Array.from(
          { length: faker.number.int({ min: 1, max: 3 }) },
          () => faker.lorem.word()
        ),
      })
    ),
    metadata: {
      createdAt: new Date("2025-01-01").toISOString(),
      updatedAt: new Date("2025-01-01").toISOString(),
      source: "faker",
    },
  };
}

describe("DiffFormatter E2E Integration", () => {
  test("should generate correct side-by-side and unified diffs for a set of changes", () => {
    faker.seed(123); // for reproducible tests
    const doc1 = createRandomUser();
    const doc2 = JSON.parse(JSON.stringify(doc1));

    doc2.username = "new-test-user";
    doc2.metadata.newProp = 12345;

    delete doc2.email;

    doc2.posts.splice(1, 1);

    doc2.posts[0].title = "A Completely New Title";

    doc2.posts.push({
      postId: "new-post-id",
      title: "A Fresh Post",
      content: "This is a brand new post added to the list.",
      tags: ["new", "exciting"],
    });

    const plan = buildPlan(userSchema, { primaryKeyMap: { "/posts": "postId" } });
    const patcher = new JsonSchemaPatcher({ plan });
    const patch = patcher.execute({original: doc1, modified: doc2});

    const formatter = new DiffFormatter(doc1, doc2);
    const sideBySideDiff = formatter.format(patch);

    expect(sideBySideDiff).toMatchSnapshot("side-by-side-diff");
  });
});
