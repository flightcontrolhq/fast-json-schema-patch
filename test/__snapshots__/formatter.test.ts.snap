// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`DiffFormatter E2E Integration should generate correct side-by-side and unified diffs for a set of changes: side-by-side-diff 1`] = `
{
  "newLines": [
    {
      "content": "{",
      "lineNumber": 1,
      "type": "unchanged",
    },
    {
      "content": "  "userId": "b438b6fa-765b-4706-8b22-88adb9b5534a",",
      "lineNumber": 2,
      "type": "unchanged",
    },
    {
      "content": "  "username": "new-test-user",",
      "lineNumber": 3,
      "type": "added",
    },
    {
      "content": "  "posts": [",
      "lineNumber": 4,
      "type": "unchanged",
    },
    {
      "content": "    {",
      "lineNumber": 5,
      "type": "unchanged",
    },
    {
      "content": "      "postId": "47f891d9-8546-4ae8-aa99-ad1c3391eab0",",
      "lineNumber": 6,
      "type": "unchanged",
    },
    {
      "content": "      "title": "A Completely New Title",",
      "lineNumber": 7,
      "type": "added",
    },
    {
      "content": "      "content": "Turba contabesco adimpleo cogo cubicularis sufficio vulnero constans tempus optio. Arcus cubitum caput coniecto dolores sophismata amiculum aperte. Solvo tyrannus ipsum ultra creo.\\nConsidero assumenda tricesimus confido ipsam. Enim abbas voveo ventosus beatus clarus eius. Vorago causa magni timor cruentus suus arto paulatim uredo vorago.\\nCursus bardus defendo. Aliquid clibanus vestigium modi defluo tardus talis. Suggero tumultus asperiores.",",
      "lineNumber": 8,
      "type": "unchanged",
    },
    {
      "content": "      "tags": [",
      "lineNumber": 9,
      "type": "unchanged",
    },
    {
      "content": "        "cito",",
      "lineNumber": 10,
      "type": "unchanged",
    },
    {
      "content": "        "cohibeo",",
      "lineNumber": 11,
      "type": "unchanged",
    },
    {
      "content": "        "sono"",
      "lineNumber": 12,
      "type": "unchanged",
    },
    {
      "content": "      ]",
      "lineNumber": 13,
      "type": "unchanged",
    },
    {
      "content": "    },",
      "lineNumber": 14,
      "type": "unchanged",
    },
    {
      "content": "    {",
      "lineNumber": 15,
      "type": "unchanged",
    },
    {
      "content": "      "postId": "134a4505-8294-4a0e-b02c-0bfdbf6d5e2f",",
      "lineNumber": 16,
      "type": "unchanged",
    },
    {
      "content": "      "title": "Anser patior acceptus supplanto absens ago caelum utique convoco.",",
      "lineNumber": 17,
      "type": "unchanged",
    },
    {
      "content": "      "content": "Minus caelum mollitia solvo coadunatio curatio defessus. Ocer vinum iure desino abduco denique vestigium avarus admoveo cum. Umquam acquiro verumtamen stabilis ventito.\\nTonsor condico consectetur cribro temeritas corporis carbo. Ver conscendo rem certus basium conduco compello vaco traho sui. Curo caries angulus coerceo arbustum aliquam pax convoco maiores aureus.\\nBibo chirographum talis laborum conculco illo suasoria verecundia. Candidus tamen tergiversatio averto tracto delego tergum. Concedo viriliter solus terebro.",",
      "lineNumber": 18,
      "type": "unchanged",
    },
    {
      "content": "      "tags": [",
      "lineNumber": 19,
      "type": "unchanged",
    },
    {
      "content": "        "bardus",",
      "lineNumber": 20,
      "type": "unchanged",
    },
    {
      "content": "        "deludo",",
      "lineNumber": 21,
      "type": "unchanged",
    },
    {
      "content": "        "tolero"",
      "lineNumber": 22,
      "type": "unchanged",
    },
    {
      "content": "      ]",
      "lineNumber": 23,
      "type": "unchanged",
    },
    {
      "content": "    },",
      "lineNumber": 24,
      "type": "unchanged",
    },
    {
      "content": "    {",
      "lineNumber": 25,
      "type": "added",
    },
    {
      "content": "      "postId": "new-post-id",",
      "lineNumber": 26,
      "type": "added",
    },
    {
      "content": "      "title": "A Fresh Post",",
      "lineNumber": 27,
      "type": "added",
    },
    {
      "content": "      "content": "This is a brand new post added to the list.",",
      "lineNumber": 28,
      "type": "added",
    },
    {
      "content": "      "tags": [",
      "lineNumber": 29,
      "type": "added",
    },
    {
      "content": "        "new",",
      "lineNumber": 30,
      "type": "added",
    },
    {
      "content": "        "exciting"",
      "lineNumber": 31,
      "type": "added",
    },
    {
      "content": "      ]",
      "lineNumber": 32,
      "type": "added",
    },
    {
      "content": "    }",
      "lineNumber": 33,
      "type": "added",
    },
    {
      "content": "  ],",
      "lineNumber": 34,
      "type": "unchanged",
    },
    {
      "content": "  "metadata": {",
      "lineNumber": 35,
      "type": "unchanged",
    },
    {
      "content": "    "createdAt": "2025-01-01T00:00:00.000Z",",
      "lineNumber": 36,
      "type": "unchanged",
    },
    {
      "content": "    "updatedAt": "2025-01-01T00:00:00.000Z",",
      "lineNumber": 37,
      "type": "unchanged",
    },
    {
      "content": "    "source": "faker",",
      "lineNumber": 38,
      "type": "unchanged",
    },
    {
      "content": "    "newProp": 12345",
      "lineNumber": 39,
      "type": "added",
    },
    {
      "content": "  }",
      "lineNumber": 40,
      "type": "unchanged",
    },
    {
      "content": "}",
      "lineNumber": 41,
      "type": "unchanged",
    },
  ],
  "originalLines": [
    {
      "content": "{",
      "lineNumber": 1,
      "type": "unchanged",
    },
    {
      "content": "  "userId": "b438b6fa-765b-4706-8b22-88adb9b5534a",",
      "lineNumber": 2,
      "type": "unchanged",
    },
    {
      "content": "  "username": "Hortense.Kunde",",
      "lineNumber": 3,
      "type": "removed",
    },
    {
      "content": "  "email": "Tyreek.Muller@yahoo.com",",
      "lineNumber": 4,
      "type": "removed",
    },
    {
      "content": "  "posts": [",
      "lineNumber": 5,
      "type": "unchanged",
    },
    {
      "content": "    {",
      "lineNumber": 6,
      "type": "unchanged",
    },
    {
      "content": "      "postId": "47f891d9-8546-4ae8-aa99-ad1c3391eab0",",
      "lineNumber": 7,
      "type": "unchanged",
    },
    {
      "content": "      "title": "Articulus argumentum sub comis stultus iste creta.",",
      "lineNumber": 8,
      "type": "removed",
    },
    {
      "content": "      "content": "Turba contabesco adimpleo cogo cubicularis sufficio vulnero constans tempus optio. Arcus cubitum caput coniecto dolores sophismata amiculum aperte. Solvo tyrannus ipsum ultra creo.\\nConsidero assumenda tricesimus confido ipsam. Enim abbas voveo ventosus beatus clarus eius. Vorago causa magni timor cruentus suus arto paulatim uredo vorago.\\nCursus bardus defendo. Aliquid clibanus vestigium modi defluo tardus talis. Suggero tumultus asperiores.",",
      "lineNumber": 9,
      "type": "unchanged",
    },
    {
      "content": "      "tags": [",
      "lineNumber": 10,
      "type": "unchanged",
    },
    {
      "content": "        "cito",",
      "lineNumber": 11,
      "type": "unchanged",
    },
    {
      "content": "        "cohibeo",",
      "lineNumber": 12,
      "type": "unchanged",
    },
    {
      "content": "        "sono"",
      "lineNumber": 13,
      "type": "unchanged",
    },
    {
      "content": "      ]",
      "lineNumber": 14,
      "type": "unchanged",
    },
    {
      "content": "    },",
      "lineNumber": 15,
      "type": "unchanged",
    },
    {
      "content": "    {",
      "lineNumber": 16,
      "type": "removed",
    },
    {
      "content": "      "postId": "1aeb77c9-19d5-4ec9-bc1d-de21266813c7",",
      "lineNumber": 17,
      "type": "removed",
    },
    {
      "content": "      "title": "Ipsam vicissitudo non.",",
      "lineNumber": 18,
      "type": "removed",
    },
    {
      "content": "      "content": "Supellex cras sordeo adamo. Adduco tamdiu demergo animus illo aeger solitudo vulnus. Natus ambulo subseco solvo admiratio textor earum cursim tertius.\\nDenique attonbitus commemoro tutis audentia cur. Capillus veritatis versus aliqua delectus desipio combibo adiuvo carbo allatus. Timor vehemens adimpleo cogito.\\nFugit quidem abscido deorsum voveo corrupti alo deinde vix congregatio. Thermae bellicus decipio sumo cunctatio aufero volaticus soleo uredo. Cenaculum desipio aegre.",",
      "lineNumber": 19,
      "type": "removed",
    },
    {
      "content": "      "tags": [",
      "lineNumber": 20,
      "type": "removed",
    },
    {
      "content": "        "capillus",",
      "lineNumber": 21,
      "type": "removed",
    },
    {
      "content": "        "corrumpo",",
      "lineNumber": 22,
      "type": "removed",
    },
    {
      "content": "        "benigne"",
      "lineNumber": 23,
      "type": "removed",
    },
    {
      "content": "      ]",
      "lineNumber": 24,
      "type": "removed",
    },
    {
      "content": "    },",
      "lineNumber": 25,
      "type": "removed",
    },
    {
      "content": "    {",
      "lineNumber": 26,
      "type": "unchanged",
    },
    {
      "content": "      "postId": "134a4505-8294-4a0e-b02c-0bfdbf6d5e2f",",
      "lineNumber": 27,
      "type": "unchanged",
    },
    {
      "content": "      "title": "Anser patior acceptus supplanto absens ago caelum utique convoco.",",
      "lineNumber": 28,
      "type": "unchanged",
    },
    {
      "content": "      "content": "Minus caelum mollitia solvo coadunatio curatio defessus. Ocer vinum iure desino abduco denique vestigium avarus admoveo cum. Umquam acquiro verumtamen stabilis ventito.\\nTonsor condico consectetur cribro temeritas corporis carbo. Ver conscendo rem certus basium conduco compello vaco traho sui. Curo caries angulus coerceo arbustum aliquam pax convoco maiores aureus.\\nBibo chirographum talis laborum conculco illo suasoria verecundia. Candidus tamen tergiversatio averto tracto delego tergum. Concedo viriliter solus terebro.",",
      "lineNumber": 29,
      "type": "unchanged",
    },
    {
      "content": "      "tags": [",
      "lineNumber": 30,
      "type": "unchanged",
    },
    {
      "content": "        "bardus",",
      "lineNumber": 31,
      "type": "unchanged",
    },
    {
      "content": "        "deludo",",
      "lineNumber": 32,
      "type": "unchanged",
    },
    {
      "content": "        "tolero"",
      "lineNumber": 33,
      "type": "unchanged",
    },
    {
      "content": "      ]",
      "lineNumber": 34,
      "type": "unchanged",
    },
    {
      "content": "    }",
      "lineNumber": 35,
      "type": "unchanged",
    },
    {
      "content": "  ],",
      "lineNumber": 36,
      "type": "unchanged",
    },
    {
      "content": "  "metadata": {",
      "lineNumber": 37,
      "type": "unchanged",
    },
    {
      "content": "    "createdAt": "2025-01-01T00:00:00.000Z",",
      "lineNumber": 38,
      "type": "unchanged",
    },
    {
      "content": "    "updatedAt": "2025-01-01T00:00:00.000Z",",
      "lineNumber": 39,
      "type": "unchanged",
    },
    {
      "content": "    "source": "faker"",
      "lineNumber": 40,
      "type": "unchanged",
    },
    {
      "content": "  }",
      "lineNumber": 41,
      "type": "unchanged",
    },
    {
      "content": "}",
      "lineNumber": 42,
      "type": "unchanged",
    },
  ],
}
`;

exports[`DiffFormatter E2E Integration should generate correct side-by-side and unified diffs for a set of changes: unified-diff 1`] = `
[
  {
    "content": "{",
    "key": "unchanged-1-1",
    "newLineNumber": 1,
    "oldLineNumber": 1,
    "type": "unchanged",
  },
  {
    "content": "  "userId": "b438b6fa-765b-4706-8b22-88adb9b5534a",",
    "key": "unchanged-2-2",
    "newLineNumber": 2,
    "oldLineNumber": 2,
    "type": "unchanged",
  },
  {
    "content": "  "username": "Hortense.Kunde",",
    "key": "removed-3",
    "oldLineNumber": 3,
    "type": "removed",
  },
  {
    "content": "  "email": "Tyreek.Muller@yahoo.com",",
    "key": "removed-4",
    "oldLineNumber": 4,
    "type": "removed",
  },
  {
    "content": "  "username": "new-test-user",",
    "key": "added-3",
    "newLineNumber": 3,
    "type": "added",
  },
  {
    "content": "  "posts": [",
    "key": "unchanged-5-4",
    "newLineNumber": 4,
    "oldLineNumber": 5,
    "type": "unchanged",
  },
  {
    "content": "    {",
    "key": "unchanged-6-5",
    "newLineNumber": 5,
    "oldLineNumber": 6,
    "type": "unchanged",
  },
  {
    "content": "      "postId": "47f891d9-8546-4ae8-aa99-ad1c3391eab0",",
    "key": "unchanged-7-6",
    "newLineNumber": 6,
    "oldLineNumber": 7,
    "type": "unchanged",
  },
  {
    "content": "      "title": "Articulus argumentum sub comis stultus iste creta.",",
    "key": "removed-8",
    "oldLineNumber": 8,
    "type": "removed",
  },
  {
    "content": "      "title": "A Completely New Title",",
    "key": "added-7",
    "newLineNumber": 7,
    "type": "added",
  },
  {
    "content": "      "content": "Turba contabesco adimpleo cogo cubicularis sufficio vulnero constans tempus optio. Arcus cubitum caput coniecto dolores sophismata amiculum aperte. Solvo tyrannus ipsum ultra creo.\\nConsidero assumenda tricesimus confido ipsam. Enim abbas voveo ventosus beatus clarus eius. Vorago causa magni timor cruentus suus arto paulatim uredo vorago.\\nCursus bardus defendo. Aliquid clibanus vestigium modi defluo tardus talis. Suggero tumultus asperiores.",",
    "key": "unchanged-9-8",
    "newLineNumber": 8,
    "oldLineNumber": 9,
    "type": "unchanged",
  },
  {
    "content": "      "tags": [",
    "key": "unchanged-10-9",
    "newLineNumber": 9,
    "oldLineNumber": 10,
    "type": "unchanged",
  },
  {
    "content": "        "cito",",
    "key": "unchanged-11-10",
    "newLineNumber": 10,
    "oldLineNumber": 11,
    "type": "unchanged",
  },
  {
    "content": "        "cohibeo",",
    "key": "unchanged-12-11",
    "newLineNumber": 11,
    "oldLineNumber": 12,
    "type": "unchanged",
  },
  {
    "content": "        "sono"",
    "key": "unchanged-13-12",
    "newLineNumber": 12,
    "oldLineNumber": 13,
    "type": "unchanged",
  },
  {
    "content": "      ]",
    "key": "unchanged-14-13",
    "newLineNumber": 13,
    "oldLineNumber": 14,
    "type": "unchanged",
  },
  {
    "content": "    },",
    "key": "unchanged-15-14",
    "newLineNumber": 14,
    "oldLineNumber": 15,
    "type": "unchanged",
  },
  {
    "content": "    {",
    "key": "removed-16",
    "oldLineNumber": 16,
    "type": "removed",
  },
  {
    "content": "      "postId": "1aeb77c9-19d5-4ec9-bc1d-de21266813c7",",
    "key": "removed-17",
    "oldLineNumber": 17,
    "type": "removed",
  },
  {
    "content": "      "title": "Ipsam vicissitudo non.",",
    "key": "removed-18",
    "oldLineNumber": 18,
    "type": "removed",
  },
  {
    "content": "      "content": "Supellex cras sordeo adamo. Adduco tamdiu demergo animus illo aeger solitudo vulnus. Natus ambulo subseco solvo admiratio textor earum cursim tertius.\\nDenique attonbitus commemoro tutis audentia cur. Capillus veritatis versus aliqua delectus desipio combibo adiuvo carbo allatus. Timor vehemens adimpleo cogito.\\nFugit quidem abscido deorsum voveo corrupti alo deinde vix congregatio. Thermae bellicus decipio sumo cunctatio aufero volaticus soleo uredo. Cenaculum desipio aegre.",",
    "key": "removed-19",
    "oldLineNumber": 19,
    "type": "removed",
  },
  {
    "content": "      "tags": [",
    "key": "removed-20",
    "oldLineNumber": 20,
    "type": "removed",
  },
  {
    "content": "        "capillus",",
    "key": "removed-21",
    "oldLineNumber": 21,
    "type": "removed",
  },
  {
    "content": "        "corrumpo",",
    "key": "removed-22",
    "oldLineNumber": 22,
    "type": "removed",
  },
  {
    "content": "        "benigne"",
    "key": "removed-23",
    "oldLineNumber": 23,
    "type": "removed",
  },
  {
    "content": "      ]",
    "key": "removed-24",
    "oldLineNumber": 24,
    "type": "removed",
  },
  {
    "content": "    },",
    "key": "removed-25",
    "oldLineNumber": 25,
    "type": "removed",
  },
  {
    "content": "    {",
    "key": "unchanged-26-15",
    "newLineNumber": 15,
    "oldLineNumber": 26,
    "type": "unchanged",
  },
  {
    "content": "      "postId": "134a4505-8294-4a0e-b02c-0bfdbf6d5e2f",",
    "key": "unchanged-27-16",
    "newLineNumber": 16,
    "oldLineNumber": 27,
    "type": "unchanged",
  },
  {
    "content": "      "title": "Anser patior acceptus supplanto absens ago caelum utique convoco.",",
    "key": "unchanged-28-17",
    "newLineNumber": 17,
    "oldLineNumber": 28,
    "type": "unchanged",
  },
  {
    "content": "      "content": "Minus caelum mollitia solvo coadunatio curatio defessus. Ocer vinum iure desino abduco denique vestigium avarus admoveo cum. Umquam acquiro verumtamen stabilis ventito.\\nTonsor condico consectetur cribro temeritas corporis carbo. Ver conscendo rem certus basium conduco compello vaco traho sui. Curo caries angulus coerceo arbustum aliquam pax convoco maiores aureus.\\nBibo chirographum talis laborum conculco illo suasoria verecundia. Candidus tamen tergiversatio averto tracto delego tergum. Concedo viriliter solus terebro.",",
    "key": "unchanged-29-18",
    "newLineNumber": 18,
    "oldLineNumber": 29,
    "type": "unchanged",
  },
  {
    "content": "      "tags": [",
    "key": "unchanged-30-19",
    "newLineNumber": 19,
    "oldLineNumber": 30,
    "type": "unchanged",
  },
  {
    "content": "        "bardus",",
    "key": "unchanged-31-20",
    "newLineNumber": 20,
    "oldLineNumber": 31,
    "type": "unchanged",
  },
  {
    "content": "        "deludo",",
    "key": "unchanged-32-21",
    "newLineNumber": 21,
    "oldLineNumber": 32,
    "type": "unchanged",
  },
  {
    "content": "        "tolero"",
    "key": "unchanged-33-22",
    "newLineNumber": 22,
    "oldLineNumber": 33,
    "type": "unchanged",
  },
  {
    "content": "      ]",
    "key": "unchanged-34-23",
    "newLineNumber": 23,
    "oldLineNumber": 34,
    "type": "unchanged",
  },
  {
    "content": "    }",
    "key": "unchanged-35-24",
    "newLineNumber": 24,
    "oldLineNumber": 35,
    "type": "unchanged",
  },
  {
    "content": "    {",
    "key": "added-25",
    "newLineNumber": 25,
    "type": "added",
  },
  {
    "content": "      "postId": "new-post-id",",
    "key": "added-26",
    "newLineNumber": 26,
    "type": "added",
  },
  {
    "content": "      "title": "A Fresh Post",",
    "key": "added-27",
    "newLineNumber": 27,
    "type": "added",
  },
  {
    "content": "      "content": "This is a brand new post added to the list.",",
    "key": "added-28",
    "newLineNumber": 28,
    "type": "added",
  },
  {
    "content": "      "tags": [",
    "key": "added-29",
    "newLineNumber": 29,
    "type": "added",
  },
  {
    "content": "        "new",",
    "key": "added-30",
    "newLineNumber": 30,
    "type": "added",
  },
  {
    "content": "        "exciting"",
    "key": "added-31",
    "newLineNumber": 31,
    "type": "added",
  },
  {
    "content": "      ]",
    "key": "added-32",
    "newLineNumber": 32,
    "type": "added",
  },
  {
    "content": "    }",
    "key": "added-33",
    "newLineNumber": 33,
    "type": "added",
  },
  {
    "content": "  ],",
    "key": "unchanged-36-34",
    "newLineNumber": 34,
    "oldLineNumber": 36,
    "type": "unchanged",
  },
  {
    "content": "  "metadata": {",
    "key": "unchanged-37-35",
    "newLineNumber": 35,
    "oldLineNumber": 37,
    "type": "unchanged",
  },
  {
    "content": "    "createdAt": "2025-01-01T00:00:00.000Z",",
    "key": "unchanged-38-36",
    "newLineNumber": 36,
    "oldLineNumber": 38,
    "type": "unchanged",
  },
  {
    "content": "    "updatedAt": "2025-01-01T00:00:00.000Z",",
    "key": "unchanged-39-37",
    "newLineNumber": 37,
    "oldLineNumber": 39,
    "type": "unchanged",
  },
  {
    "content": "    "source": "faker"",
    "key": "unchanged-40-38",
    "newLineNumber": 38,
    "oldLineNumber": 40,
    "type": "unchanged",
  },
  {
    "content": "    "newProp": 12345",
    "key": "added-39",
    "newLineNumber": 39,
    "type": "added",
  },
  {
    "content": "  }",
    "key": "unchanged-41-40",
    "newLineNumber": 40,
    "oldLineNumber": 41,
    "type": "unchanged",
  },
  {
    "content": "}",
    "key": "unchanged-42-41",
    "newLineNumber": 41,
    "oldLineNumber": 42,
    "type": "unchanged",
  },
]
`;
